improver/psychrometric_calculations/hail_size.py

```diff
@@ -0,0 +1,338 @@
+
+    def dry_adiabatic_drop_to_ccl(
+        self, ccl_pressure, temperature_at_268, pressure_at_268
+    ):
+
+        """Used to calulate c. Does a dry adiabatic descent from B to the ccl_pressure"""
+
+        t_dry = dry_adiabatic_temperature(
+            temperature_at_268.data, pressure_at_268.data, ccl_pressure.data
+        )
+        ccl_level_temperature = temperature_at_268.copy(data=t_dry)
+        ccl_level_temperature.rename("potential temperature of 268K at ccl pressure")
+        return ccl_level_temperature
+
+    def get_hail_size(self, vertical, horizontal):
+
+        """Uses the lookup_table and the vertical and horizontal indexes calculated 
+        to extract and store values from nomogram """
+
+        lookup_table = self.nomogram_values()
+        shape = np.shape(vertical)
+
+        horizontal = np.around(horizontal.data / 5, decimals=0) - 1
+        vertical = np.around(vertical.data * 2, decimals=0) - 1
+
+        horizontal_flat = horizontal.flatten(order="C")
+        vertical_flat = vertical.flatten(order="C")
+
+        hail_size_list = []
+        for vertical, horizontal in zip(vertical_flat, horizontal_flat):
+
+            if min(horizontal, vertical) < 0:
+                hail_size_list.append(0)
+            else:
+
+                if vertical > len(lookup_table) - 1:
+                    vertical = len(lookup_table) - 1
+                if horizontal > len(lookup_table[0]) - 1:
+                    horizontal = len(lookup_table[0]) - 1
+
+                hail_size_list.append(lookup_table[int(vertical)][int(horizontal)])
+
+        hail_size = np.reshape(hail_size_list, shape, order="C")
+        return hail_size
+
+    def make_hail_cube(self, hail_size, ccl_temperature, ccl_pressure):
+
+        """puts the data into a cube containing the hail_data"""
+
+        attributes = {}
+        if self.model_id_attr:
+            attributes[self.model_id_attr] = self.ccl_temperature.attributes[
+                self.model_id_attr
+            ]
+
+        hail_size_cube = create_new_diagnostic_cube(
+            name="size_of_hail_stones",
+            units="mm",
+            template_cube=ccl_temperature,
+            data=hail_size,
+            mandatory_attributes=generate_mandatory_attributes(
+                [ccl_temperature, ccl_pressure]
+            ),
+            optional_attributes=attributes,
+        )
+
+        return hail_size_cube
+
+    def check_cubes(
+        self,
+        ccl_temperature,
+        ccl_pressure,
+        temperature_on_pressure,
+        humidity_mixing_ratio_on_pressure,
+    ):
+        """function to check size and units of input cubes"""
+
+        temp_slice = next(temperature_on_pressure.slices_over("pressure"))
+
+        assert_spatial_coords_match([ccl_temperature, ccl_pressure, temp_slice])
+        assert_spatial_coords_match(
+            [temperature_on_pressure, humidity_mixing_ratio_on_pressure]
+        )
```

Comment:
For completeness, you should probably check that the horizontal coords of temperature_on_pressure match the coords of ccl_temperature.